### ISSUE TITLE
Implement MultiAgentEnvironment adapter for Bucket Brigade

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/thrust#6]
## Status

\`envs/bucket-brigade/\` is checked in as a git submodule pointing at the BB repo, but **no glue code** exists to expose \`bucket_brigade_core::BucketBrigade\` as a thrust \`MultiAgentEnvironment\`. The directory just contains the BB source tree.

## What's needed

A new module / crate (e.g. \`thrust-rl-bucket-brigade\`) implementing:

\`\`\`rust
use crate::multi_agent::environment::{MultiAgentEnvironment, MultiAgentResult};
use bucket_brigade_core::{BucketBrigade, Scenario};

pub struct BucketBrigadeMaEnv {
    inner: BucketBrigade,
    scenario: Scenario,
    num_agents: usize,
}

impl Environment for BucketBrigadeMaEnv { ... }
impl MultiAgentEnvironment for BucketBrigadeMaEnv {
    fn step_multi(&mut self, actions: &[Vec<i64>]) -> MultiAgentResult { ... }
}
\`\`\`

Observation layout should mirror the Python flattener at \`bucket_brigade/training/observation_utils.py\`:
\`[houses(10), signals(N), locations(N), last_actions(2N), scenario_info(12)]\`.

## Blockers

- #3 (multi-discrete actions in MA trait): BB uses \`[house, mode]\` per agent.
- The Rust core (\`bucket-brigade-core\`) currently exposes a higher-level \`PyVectorEnv\` and per-agent \`AgentObservation\` for Python; we want to use the lower-level engine types directly here to avoid double translation.

## Use case

Required for the Slepian-Wolf MARL P3 specialization experiment to run on thrust. Currently the experiment runs in pure Python with PyTorch (\`bucket_brigade.training.joint_trainer\`); we want a 3-6x speedup via the thrust port.

## Suggested location

Probably a separate crate under the workspace rather than \`src/env/games/bucket_brigade/\`, because BB has its own non-trivial build chain (Rust core + Python bindings + WASM target) and we shouldn't tangle thrust's build with that.

**Comment by rjwalters** [comment]
## Curator Enhancement (2026-05-14)

### Current Status (verified against `master` @ `1ddbbf5`)

**Partially implemented.** The struct, observation flattener, and step/reset surface area all exist; the trait impl is the only remaining piece, and it is blocked on #3.

What is already present on `master`:

- `src/env/games/bucket_brigade.rs` — `BucketBrigadeMaEnv` struct (153 lines) with:
  - `new(scenario, num_agents, seed)`
  - `reset(seed) -> Vec<Vec<f32>>`
  - `step(&[[u8; 2]]) -> MaStepResult` (factored `[house, mode]` actions)
  - `obs_dim()`, `action_dims() -> [10, 2]`, `num_agents()`, `scenario()`
  - `flatten_observation()` matching the Python layout `[houses(10), signals(N), locations(N), last_actions(2N), scenario_info(12)]`
- `src/env/games/mod.rs` — feature-gated `pub mod bucket_brigade` + re-export of `BucketBrigadeMaEnv`
- `Cargo.toml` — `env-bucket-brigade` feature, optional `bucket-brigade-core` path dep
- `tests/test_bucket_brigade_env.rs` — 5 passing tests (obs-dim contract, reset shape, step shape, determinism, reset-after-done regression)
- `examples/games/bucket_brigade/train_p3.rs` — already consuming `BucketBrigadeMaEnv` directly (does NOT go through the `MultiAgentEnvironment` trait)

What is explicitly **NOT** present (and why):

The file's module doc-comment (lines 17-20) and commit `273bbd5` both state:

> Does NOT implement `crate::multi_agent::MultiAgentEnvironment` because that trait currently takes `&[i64]` and doesn't support factored action spaces (see upstream issue #3). When the trait is widened, this struct can grow a trivial impl.

Confirmed in `src/multi_agent/environment.rs:36`: `fn step_multi(&mut self, actions: &[i64]) -> MultiAgentResult;` — single-discrete actions only. Bucket Brigade needs `[house_index, mode]` (multi-discrete `[10, 2]`), which cannot be expressed without either packing into one i64 (cardinality blow-up, loses factored structure) or widening the trait.

### Root Cause / Blocking Relationship

This issue is **blocked on #3** (`MultiAgentEnvironment: support multi-discrete action spaces`). The non-trait surface area has already shipped under issue #6's banner (commit `273bbd5` was committed with `Refs: #6`), but #6 as originally written asks for an `impl MultiAgentEnvironment for BucketBrigadeMaEnv`, which #3 must land first.

The joint trainer (`examples/games/bucket_brigade/train_p3.rs`) currently bypasses the trait entirely and calls the inherent `step`/`reset` methods on the concrete struct. That works for the P3 experiment but means `BucketBrigadeMaEnv` cannot be used with the generic multi-agent infrastructure (`GameSimulator`, `PolicyLearner`, etc.) until the trait impl lands.

### Recommended Disposition

Since the original ask in #6 is now **two distinct deliverables** with different unblock states, this issue should be re-scoped to its remaining work:

**Scope of remaining work for #6**: Once #3 lands and the trait signature becomes `step_multi(&mut self, actions: &[Vec<i64>])`, add a trivial `impl MultiAgentEnvironment for BucketBrigadeMaEnv` that:

1. Implements `Environment` (currently no `Environment` impl exists — required by `MultiAgentEnvironment: Environment`).
2. Implements the widened `MultiAgentEnvironment` by adapting `&[Vec<i64>]` to `&[[u8; 2]]` and delegating to the existing inherent `step`.
3. Maps `MaStepResult` → `MultiAgentResult`:
   - `observations` → `observations`
   - `rewards` → `rewards`
   - `done` → broadcast to `terminated: vec![done; num_agents]`, `truncated: vec![false; num_agents]`
4. Adds `get_agent_observation(usize) -> Vec<f32>` (already easy via `flatten_observation(self.inner.get_observation(i))`).
5. Adds `active_agents()` — for BB, all agents remain active for the whole episode (the game ends globally), so this is `vec![true; self.num_agents]`.
6. Adds `agent_action_space(i) -> Vec<usize>` returning `vec![10, 2]` (the exact shape will follow whatever #3 settles on).

### Implementation Guidance

**Do NOT** start work on this until #3 has merged. The trait signature will be settled by #3, and any impl written against the current `&[i64]` signature would either need to pack the factored action or be discarded.

**Once #3 has merged**, the impl is straightforward (~40-60 LOC):

```rust
// Sketch — exact signatures depend on the resolution of #3.
use crate::env::{Environment, SpaceInfo, SpaceType, StepInfo, StepResult};
use crate::multi_agent::environment::{MultiAgentEnvironment, MultiAgentResult};

impl Environment for BucketBrigadeMaEnv {
    type Observation = Vec<f32>;     // flattened agent-0 obs by convention
    type Action = Vec<i64>;          // [house, mode] for agent 0 in single-agent fallback
    // ... reset/step/observation_space/action_space ...
}

impl MultiAgentEnvironment for BucketBrigadeMaEnv {
    fn num_agents(&self) -> usize { self.num_agents }

    fn get_agent_observation(&self, agent_id: usize) -> Vec<f32> {
        flatten_observation(&self.inner.get_observation(agent_id))
    }

    fn step_multi(&mut self, actions: &[Vec<i64>]) -> MultiAgentResult {
        let packed: Vec<[u8; 2]> = actions.iter()
            .map(|a| [a[0] as u8, a[1] as u8])
            .collect();
        let r = self.step(&packed);
        MultiAgentResult::new(
            r.observations,
            r.rewards,
            vec![r.done; self.num_agents],   // BB terminates globally
            vec![false; self.num_agents],
        )
    }

    fn active_agents(&self) -> Vec<bool> {
        vec![true; self.num_agents]          // BB has no per-agent termination
    }
}
```

The `flatten_observation` helper is already package-private in the same file. The action-shape check inside `step_multi` should `debug_assert_eq!(actions[i].len(), 2)` to catch misconfigured policies early.

### Acceptance Criteria

- [ ] #3 has merged (widened `step_multi` signature is on `master`).
- [ ] `BucketBrigadeMaEnv` implements `crate::env::Environment` (prerequisite for `MultiAgentEnvironment`).
- [ ] `BucketBrigadeMaEnv` implements `crate::multi_agent::MultiAgentEnvironment` with the new signature.
- [ ] All 5 existing tests in `tests/test_bucket_brigade_env.rs` still pass.
- [ ] At least one new test exercises the trait method (`step_multi` with `&[Vec<i64>]`) and asserts:
  - returned `MultiAgentResult.observations.len() == num_agents`
  - returned `MultiAgentResult.rewards.len() == num_agents`
  - `terminated.len() == num_agents`, all equal (BB is global-termination)
  - `active_agents().iter().all(|&a| a)` while episode is live
- [ ] `train_p3.rs` either continues to compile unchanged (calls inherent methods) **or** is migrated to use the trait — pick one explicitly, do not leave the example half-migrated.
- [ ] Module doc-comment at `src/env/games/bucket_brigade.rs:17-20` is updated/removed (it currently states the trait is NOT implemented).
- [ ] `cargo build --features env-bucket-brigade` and `cargo build --features 'env-bucket-brigade training'` both succeed.

### Affected Files

- `src/env/games/bucket_brigade.rs` — Add `impl Environment` and `impl MultiAgentEnvironment` blocks; update module doc-comment lines 17-20.
- `tests/test_bucket_brigade_env.rs` — Add at least one test going through the `MultiAgentEnvironment` trait (not just inherent methods).
- `examples/games/bucket_brigade/train_p3.rs` — No change required; optional migration to use the trait surface instead of inherent methods (the inherent API gives slightly tighter types — `[[u8; 2]]` — so keeping the direct path is reasonable for the hot training loop).
- `src/multi_agent/environment.rs` — Read-only reference; modified by #3, not this issue.

### Test Plan

- [ ] Manual verification: `cargo test --test test_bucket_brigade_env --features env-bucket-brigade` — all existing + new trait-level tests pass.
- [ ] Automated tests: New `step_multi_trait_returns_per_agent_results` test in `tests/test_bucket_brigade_env.rs` exercising the trait method end-to-end.
- [ ] Automated tests: New `multi_agent_trait_object_compiles` test or doc-test confirming `BucketBrigadeMaEnv` can be coerced to `dyn MultiAgentEnvironment` (or held as `Box<dyn MultiAgentEnvironment>`), since this is the practical reason to land the trait impl.
- [ ] Regression: Existing P3 trainer (`cargo run --example train_p3 --features 'training env-bucket-brigade' -- --num-agents 4 --num-steps 32`) still trains for at least one update without panicking.
- [ ] Edge case: `step_multi` with wrong-length `actions[i]` (e.g., `vec![0i64]` instead of `vec![0, 0]`) panics with a clear error or returns a typed error — pick one consistent with how `MultiAgentEnvironment` handles input validation elsewhere in the codebase.

### Dependencies

- [ ] #3: `MultiAgentEnvironment: support multi-discrete action spaces` — this issue cannot proceed until the trait signature widens. The pre-requisite relationship is already noted in both #3 and #6's body.

This issue must remain `loom:blocked` until #3 closes. When #3's PR merges to `master`, the `loom:blocked` label can be removed and the issue is ready for human approval (`loom:issue`).

### Notes for Builder

- The work is genuinely small (~40-60 LOC of trait impl + ~30 LOC of tests). The complexity is entirely in waiting for #3.
- Do not refactor the inherent `step`/`reset` methods — `train_p3.rs` depends on the tighter `[[u8; 2]]` types and removing them would be a separate, larger change.
- The `BucketBrigade::reset()` API on `bucket-brigade-core` does not currently accept a seed; the inherent `reset(Option<u64>)` works around this by re-`new`-ing the engine when a seed is supplied. The trait impl should *not* re-expose this distinction — its `Environment::reset` will just call the seedless path.

**Comment by rjwalters** [comment]
Closing as completed — implemented as part of PR #14 (commit 602d32c).

`impl MultiAgentEnvironment for BucketBrigadeMaEnv` is now in `src/env/games/bucket_brigade.rs:188` with all five trait methods (`num_agents`, `agent_action_space`, `step_multi`, `reset_multi`, `get_agent_observation`). The FIXME at the top of the file from commit 273bbd5 has been removed.

Integration coverage: `tests/test_multi_discrete_env.rs` was added in #14 and exercises the multi-discrete trait path on BucketBrigade.
